### RELAX ON CYP8B1 gene in 7 Afrotheria species.

Here is the script of running RELAX on CYP8B1 gene with 7 Afrotheria species. Each species is taken as a forground (fg) species with rest of the 6 species as background (bg).
Here is the wrkflow :

#### Here is the full workflow :

<p align="center">
  <img width="840" height="150" src="https://iailga.dm.files.1drv.com/y4m8k1sWndFc2_TiZ8hJW0sdeaNDIOosc707XJQ0jghyjjLBkRGJj70Bdi57erxhF-29XCkRwfBMfs1Wu_QJNxwbzSREdtdZ8WVIvKShb1Di6NQEvc3ULrgCjQCHmH-naNxXNL-5Y-AxEia1TFdjhw5H5KwNgFl6pctUPpaYBZF7vL-5WSPqUH3CMzHCptYv8Dx86JFXiK25mSSc-DV516L4Q?width=1232&height=202&cropmode=none">
</p>

#### 1. Download ORFs

    cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
    wget https://raw.githubusercontent.com/ceglab/CYP8B1/master/ORFs/2_Afrotheria.fa
    mv 2_Afrotheria.fa afrotheria.fa

#### 2. Introduce Gaps
    cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
    bioawk -cfastx '{print $name, length($seq)}' afrotheria.fa > gene_length.txt
    bedtools random -n 3 -l 5 -g gene_length.txt -seed 666 > l5_n3_gaps.bed
    bedtools maskfasta -fi afrotheria.fa -bed l5_n3_gaps.bed -fo masked.fa

#### 3. Run MSA with guidance
    perl /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/tools/guidance.v2.02/www/Guidance/guidance.pl --program
    GUIDANCE --seqFile masked.fa --msaProgram PRANK --seqType codon --outDir prank_100 --genCode 1 --bootstraps 100 --proc_num 8
    seqkit sort prank_100/MSA.PRANK.aln.With_Names > CDM_PRANK.aln

#### 4. Find best model.
    modeltest-ng -i CDM_PRANK.aln -t ml -T raxml

#### 5. Infer best tree
    raxml-ng --all --msa CDM_PRANK.aln --model HKY+G4 --bs-trees 1000 --threads 2

    

#### 6. Create species list & Label tree

    t=`grep ">" CDM_PRANK.aln|wc -l`
    grep ">" CDM_PRANK.aln|sed 's/>//g' > taxlist.txt
    
    for i in `cat taxlist.txt`
    do sed "s/$i/$i{fg}/g" CDM_PRANK.aln.raxml.bestTree > "$i"_treeLabled.txt
    done
    
    
 <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary>: - :point_right: What is sed command doing ?</summary>
    
   <div style="float: right; margin-left: 30px;"><img title="How sed works" style="float: right;margin-left: 30px;" src="https://fkgfrg.dm.files.1drv.com/y4ms8I6BwbFAsMtKzo6QB82f-T3-ZSdl7Deral6o8Zkb81hF7gamLTCy6n0z5uGMqKG6WIS8WGacVh3XUFCeXfD74FQwAi0qMIwfYL1tqES68iueTY-ajB7OsbLpi4bEcnfkTsstJPp-6ImBEQEqHNLoAMuNv4GXF8T4T9Kqt3wrwocCfbmlBtclpjYMFXd10Tc53MhQizw-5D_YqtHh8Eqcg?width=1513&height=705&cropmode=none" align=right height=220/></div>
     
    
   - SED command in UNIX is stands for stream editor.
   - Functions - search, find and replace, insert or delete inside a file.
   - Here we have to label the tree with one species as forground species by putting 'fg' in the bracket.
    
</details>
<!-- AUTO-GENERATED-CONTENT:END -->

#### 7. Make config file for HYPHY
    for i in `cat taxlist.txt`
    do echo -ne  "14\n9\n1\n"$PWD"/CDM_PRANK.aln\n"$PWD"/"$i"_treeLabled.txt\n2\n2" > "$i"_tree.config
    done

#### 8. Running Hyphy minimal
    for i in `cat taxlist.txt`
    do HYPHYMP<"$i"_tree.config > "$i"_treeoutput_HYPHY
    done

#### 9. Check the results

    for i in $(ls *_HYPHY);do echo $i; grep "Relaxation/intensification parameter (K)" $i; tail -4 $i ;done

    for i in $(ls *_HYPHY);do grep "Relaxation/intensification parameter (K)" $i ;done
    for i in $(ls *_HYPHY);do grep "Likelihood ratio test" $i ;done

### All output files of step 2 to step 8 :

<p align="center">
  <img width="1200" height="500" src="https://hkhoua.dm.files.1drv.com/y4mbjIxMQLQFKmB8roesXyzOaixKsEWobVMhrNtVaG6BXj4CCVR52pAKZr6NEiwc44pRRH1lX1VCzTdJ5KmU1kbT1VVZJxNwwb67PM5nyfgxHaRq75jVcecaUrz3UMcAYw7tnAhHc8fBDBNiKv8lZ7VJrUjUW6YA_vtEURvpU63KcdNt7VMVHFgVh5oFIbcWezb_Fh9Kmrh2CGME65jfPNH1w?width=1879&height=950&cropmode=none">
</p>
