### Objective - To analyze how the detection of relaxed selection of CYP8B1 gene by `RELAX` is affected by the following :
- number of gaps
- length of gaps 
- number of species


### Steps - MSA's and Tree Building

- **1. Select the order and number of species to be used.**  
  
   <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> The order Afrotheria can be used - because they show strong relaxed selection on CYP8B1 gene in Shinde 2019.    :point_right: Points to note : </summary>

  ---

    - There were only 7 Afrothrtia species used in Shinde et al.,'s study.
    - Gaps 'replace' random nucleotides with N's (not 'insert' - which will change reading frame).
    - Even before introducing gaps - there were total 16 N's already present in the sequence of 2 species.
      - 13 N's in Loxodonta africana (4 triplets + 1 nt)
      - 3 N's in Tricheus manatus latirostris (1 triplet)
    - These N's are also gaps - but are introduced before ORF validation to facilitate alignment.
    - Hence these N's are considered normal w.r.t **gap N's** that we introduce in this step.
     
    ---  
  
     </details>
     <!-- AUTO-GENERATED-CONTENT:END -->

---

- **2. Introduce the gaps using bedtools**

  - Calculate gene length of all species : output of this step is the input for the next step - `bedtools random`.
  
        cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
        bioawk -cfastx '{print $name, length($seq)}' afrotheria.fa > gene_length.txt

  - Make the intervals of where to introduce the gaps :
    - n = number of gaps.
    - l = length of gaps.
  
          bedtools random -n 3 -l 5 -g gene_length.txt -seed 666 > l5_n3_gaps.bed
 
  - Introduce the gaps using the intervals made in previous step :

        bedtools maskfasta -fi afrotheria.fa -bed l5_n3_gaps.bed -fo masked.fa

---

- **3. Run MSA using Guidance** 

  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> :point_right: What is Guidance & Why to use it ?</summary>
         
   ---      
         
   - GUIDANCE is  used for weighting, filtering or masking unreliably aligned positions in sequence alignments before subsequent analysis.
     
     <div style="float: right; margin-left: 30px;"><img title="How Guidance works" style="float: right;margin-left: 30px;" src="http://guidance.tau.ac.il/ver2/fig_method.server-1.png" align=right height=320/></div>
     
   - It points to sequences that are unreliably aligned and enables their automatic removal from the MSA, in preparation for downstream analyses.
   - GUIDANCE cannot tell you which alignment is better.
   - GUIDANCE measures the robustness of the alignment and not accuracy.
   
      - So a perfect score means - the aligner will always consistently aligns the sequences in the same way, regardless of perturbations in GUIDANCE.
      - Also a higher score doesn't mean higher accurcy.
  
  ---
  
  </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
  
      cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
      i=/home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf/afrotheria.fa
      guidance --program GUIDANCE --seqFile "$i" --msaProgram PRANK --seqType codon --outDir "$i".100_PRANK --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram MUSCLE --seqType codon --outDir "$i".100_MUSCLE --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram CLUSTALW --seqType codon --outDir "$i".100_CLUSTALW --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram MAFFT --seqType codon --outDir "$i".100_MAFFT --genCode 1 --bootstraps 100 --proc_num 8
 
 
  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary> :point_right: Error log : </summary>

     ---

  - Running `Guidance` with each aligner throwed this error :
     
     <p align="center">
     <img width="760" height="160" src="https://iqhvlw.dm.files.1drv.com/y4m0tUgOy2fNYW7OR7XZ5JxwFxJL3FbtZMvox4VR2UP5deRgZYnLjuWSvF99ETN3vrWTLxziERpOYKHu6rcSeXlYenSrEot-ZisC_UZ37e23gadCzJDgdOcAL4j3oBg_aVzlSoWGGSh-JstZgVR1PKquVZoCyzC4hsHiBVzBWR1FVSRBpa7p3FNvQWfXUngHPkeBI_6dutBJhd6Lqnsuo42WA?width=950&height=221&cropmode=none">
      </p>
  
  - The command worked fine even with this error.
  - This is not really an error that will affect the results. 
  - It might be some bug in the program.
  - However to confirm check one of the output it made - `Seqs.Codes` 
    - `Seq.Codes` is just list of sequences.
    - Make sure no sequences are repeated.      
      
   ---
      
   </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
      
--- 
 
- **4. Sort MSA files before comparing - Using github repository `seqkit` of shensei356.**
   - Sorting is done because - MSA comparison have to run in the same order in which the species are arranged in MSA.
  
         seqkit sort "$i".100_PRANK/MSA.PRANK.aln.With_Names >"$i"_PRANK.aln
         seqkit sort "$i".100_MUSCLE/MSA.MUSCLE.aln.With_Names >"$i"_MUSCLE.aln
         seqkit sort "$i".100_CLUSTALW/MSA.CLUSTALW.aln.With_Names >"$i"_CLUSTALW.aln
         seqkit sort "$i".100_MAFFT/MSA.MAFFT.aln.With_Names >"$i"_MAFFT.aln
        
        
    <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary> :point_right: Seqkit Output : </summary>

    ---

     <div style="float: right; margin-left: 30px;"><img title="Logo by @gregcaporaso." style="float: right;margin-left: 30px;" src="https://hkjwcw.dm.files.1drv.com/y4ml8T7vybLkXBQovFtyNmoXdoOA-0QH84O_kshlmiMDeMX_ejw3UdPx3R7P5NrqkFmR1mIhifKkd57PtGAn6O_AvHVtYotU52jBlp_OhfzMasQZKVZnEUkJ5_DOoZMEY8uIAxROsXiMds_lqG76CO0-OqPG30hoEImKetfTmWi7XP1fh1YvAzsTQV7auPbodIA4XYuh5gYfpY-aBTKNtzzcQ?width=928&height=327&cropmode=none" align=left height=220/></div>     

 
    &nbsp;
  
    - Above : MSA before sorting.
  
  &nbsp;
  
  &nbsp;
  
    - Below : MSA after sorting.
      
    &nbsp;
    
    &nbsp;

    
  ---
  
   </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
  
---  
  
- **5. Compare MSA's - Using `mumsa` from [msa.sbc.su.se](http://msa.sbc.su.se/)**

  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary> :point_right: What is Mumsa & why to use it ? : </summary>

    ---

  - Mumsa assess the quality of multiple sequence alignments using 2 types of scores :
    - AOS - Average Overlap Score 
      - Determining alignment 'difficulty'.
      - AOS score above 0.8 - easy to align - good agreement among the input alignments -> can be trusted.
      - AOS score below 0.5 - very difficult to align -> have to be treated with care.
       
    - MOS - Multiple Overlap Score 
      - Determining alignment 'quality'.
      - MOS score above 0.8 may be considered reliable.
      
  - Both scores are important and have to be considered jointly, because :
    - the best alignment according to the MOS score is probably not accurate enough for further studies if the AOS score for the entire alignment case is very low.

   ---

  </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
  
      echo "MUMSA comparison results of mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln" > "$i".log
      mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln >> "$i".log 
      alncomp=`mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln`
     
     &nbsp;      
   - The `i.log` file is used to store the results as we proceed, to use it later.       
     &nbsp;      
  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
    <details>
    <summary> :point_right: MUMSA Output : </summary> 
   
    ---
 
     - First number is AOS score. 
     
          <div style="float: right; margin-left: 30px;"><img title="Logo by @gregcaporaso." style="float: right;margin-left: 30px;" src="https://hqjwcw.dm.files.1drv.com/y4mBDDX-FJeH-akhh-IaHRlY9fEQmhMgjYAHHwFI1DQcLIKv8Xr2g1xQOqctX1sm3wSyLk81luvT1_YJWU2LsHQ3ucKzeJGcM94qrMQdB3QI1Zpk_vUPIOQXn_0PWcxeaxpWeDhM5pU-EWiNYVz_ljOrH6Peb6tgP2DLdlKqS0WgSYIIkol3T2_qlI8H4mIWl9HLsrLsqo3_Iaoazyy_WtnmA?width=811&height=56&cropmode=none" align=right height=40/></div>

     - Rest four are MOS scores for different aligners.
       - Since all aligners have same score - not mentioning which score belongs to which aligner.
 
    ---
 
   </details>
   <!-- AUTO-GENERATED-CONTENT:END -->
   
---   
      
- **6. Perform model testing on each alignment - Using github repository `modeltest` of ddarriba.**

    <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
    <details>
    <summary> :point_right: What is modeltest-ng and why to use it ? : </summary> 
 
   ---
 
    - Modeltest-ng is a tool for selecting the best-fit model of evolution for DNA and protein alignments.
    - It involves finding that evolutionary tree which yields the highest probability of evolving the observed data.
    - Here we used Modeltest-ng to find the model best suited for our MSA's.

   ---

    </details>
    <!-- AUTO-GENERATED-CONTENT:END -->

      modeltest-ng -i "$i"_MAFFT.aln -t ml -T raxml
      modeltest-ng -i "$i"_PRANK.aln -t ml -T raxml
      modeltest-ng -i "$i"_MUSCLE.aln -t ml -T raxml
      modeltest-ng -i "$i"_CLUSTALW.aln -t ml -T raxml
      
  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
    <details>
    <summary> :point_right: ModelTest_NG Output : </summary> 
   
    ---
   
     - For each alignment - it gives the best model according to 3 criterias - BIC, AIC snd AICc.
     - Here the best model according all three is `HKY+G4`.
    
     <p align="center">
     <img width="850" height="300" src="https://e6jwcw.dm.files.1drv.com/y4mf_3ELIK23zZAVOJSc9pgmmr6wLdYkBLJVV5WwS2wtD2mxy-FHOMLOnyC47s31H5yeNOtuKaV0oe58QOhaJEym0G2OoPH5oZqWMDVx2wuCJXZecRcQ6qXVdOCBNHs3-Dhk0m09CPZZEsATbCM7c27F5Oc--JJDbDneV3vg2gKCI6dOh0yobgVW6fIdEwjumybU5DylEFiTp-oqxXsBkY4ng?width=2010&height=591&cropmode=none">
      </p>
 
    ---
 
    </details>
    <!-- AUTO-GENERATED-CONTENT:END -->  
      
   - Check which is the best model according to BIC and AICc and save them in varaibles.   
 
         for j in "$i"_PRANK "$i"_MUSCLE "$i"_CLUSTALW "$i"_MAFFT
         do
         BICmodel=`grep -A 2 "Best model according to BIC" "$j".aln.out|tail -1|awk '{print $2}'`
         AICmodel=`grep -A 2 "Best model according to AICc" "$j".aln.out|tail -1|awk '{print $2}'`
         echo $j $BICmodel $AICmodel
         done

   - Choose an alignment used & save the corrsponding best model into a variable called `model`.
   
         j="$i"_PRANK
         model=`grep -A 2 "Best model according to BIC" "$j".aln.out|tail -1|awk '{print $2}'`
 
   - Add this result to the log file.
       
         echo "Model $model has been selected." >> "$i".log

---

- **7. Build phylogenies with bootstrap - Using github repository `raxmlng` of amkozlov**

   <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
    <details>
    <summary> :point_right: What is RAxML-NG & why to use it ?: </summary> 
   
   ---
   
     - RAxML-NG is a phylogenetic tree inference tool which uses maximum-likelihood (ML) optimality criterion.
     - RAxML-NG is a successor of RAxML (Stamatakis 2014) which leverages the highly optimized likelihood computation.
     - Here we used RAxML-NG to infer the best trees using the best model we identified in the previous step.
 
   ---
 
    </details>
    <!-- AUTO-GENERATED-CONTENT:END -->  
    
   - Before starting check whether the MSA doesn't contain sites with undetermined characters  or duplicate taxon names, etc. 
   - This is super-important as more than 50% of all failed RAxML runs are due to tree or MSA format errors!.
   
         raxml-ng --check --msa afrotheria.fa_PRANK.aln --model $model
         
   - Perform an all-in-one analysis (ML tree search + non-parametric bootstrap) with 1000 bootstrap replicates.
     - Use `time` command to know command ececution time.
     - Use less number of threads for optimum throughput.
     - Save the root taxa in a variable called `roottax`.
   
           time raxml-ng --all -msa "$i"_PRANK.aln --model $model --bs-trees 1000 --threads 2
           roottax=`grep ">" "$i"_PRANK.aln|head -1|sed 's/>//g'`
           
  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
    <details>
    <summary> :point_right: RAxML-NG Output : </summary> 
  
  ---
   
     - It produced 8 files :
       - Top panel - the total 8 files it produced.
       - Bottom panel - 5 important files out of 8 files and what are they.
   
   <p align="center">
   <img width="890" height="190" src="https://fkjwcw.dm.files.1drv.com/y4mP-fRnW7AAHPEkZNjAyKd3bgah5IazzHed4GFvAoQV7JYkLDHHxwm5e_bWNIiPl1Dnj3PPejjSJi4kp6sAziGklf8t-8964js2JPFh-RDrDOHC_xXliTCi-EIAN5Fb9DQWW8STYrMXk8J4YJubRaXezKd1CDk6a8W88KDZ9M-0X4VDhWi-wWNrav-x46N7x-ae_RkSpBIXlym1rGs9TQ9_g?width=1958&height=369&cropmode=none">
   </p>
   
  ---   
 
    </details>
    <!-- AUTO-GENERATED-CONTENT:END -->    
    
---    
           
- **8. Reroot and sort trees before counting topologies using [newick Utilities](http://cegg.unige.ch/newick_utils)**

  - Find how many topologies are supported by more than 10 percent and 1 percent of the total bootstrapped trees and save it in a variable.
  
     <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
     <details>
     <summary> :point_right: What is Newick Utilities  & why to use it ?: </summary> 
  
    ---
  
     - The Newick Utilities are a suite of Unix shell tools for processing phylogenetic trees.
     - Functions include :
       - re-rooting 
       - extracting subtrees 
       - trimming 
       - pruning 
       - condensing 
       - drawing
     - Each of these functions are have seperate independent command or script to execute them. 
     - Here we used to extract, reroot and order and the trees.
     - Finally, find how many topologies are supported by more than 10 percent and 1 percent of the trees.
     
     ---  
     
     </details>
    <!-- AUTO-GENERATED-CONTENT:END -->  
    
  - For more than 10 percent (i.e., 100) :  

        nw_topology "$i"_PRANK.aln.raxml.bootstraps|nw_reroot - $roottax|nw_order -|sort|uniq -c|awk '$1>100{print $2}' > 10per_more.txt
        top10per=`cat 10per_more.txt|wc -l`
        echo "$top10per topologies supported by more than 10 percent of the trees." >> "$i".log
      
   - For more than 1 percent (i.e., 10) :
      
         nw_topology "$i"_PRANK.aln.raxml.bootstraps|nw_reroot - $roottax|nw_order -|sort|uniq -c|awk '$1>10{print $2}' > 1per_more.txt
         top1per=`cat 1per_more.txt|wc -l`
         echo "$top1per topologies supported by more than 1 percent of the trees." >> "$i".log
         
---

- **9. Save some info in variables and files** 
     - taxcount - number of taxas (species) used for MSA.
     - taxlist - List of taxas used in the MSA.
   
           taxcount=`grep ">" "$i"_PRANK.aln|wc -l`
           grep ">" "$i"_PRANK.aln|sed 's/>//g' > taxlist.txt
     
---     
--- 
 
- **10. Save the main results of all these step in a file in the preceeding folder.**           

      echo -ne "$i\t$alncomp\t$model\t$roottax\t$top10per\t$top1per\n" >> ../all.stats
         
### All output files made in this 10 steps :

<p align="center">
  <img width="960" height="300" src="https://g6ilga.dm.files.1drv.com/y4mRQW-diCVHP-TUYKJL9nkhez_eScxj-B8n1FZJ-UO9mG8gp2pczTCJ8mV1on0J9jejaN2zjfgvafI9qWKmxrDEHlAOdebRvXQ9vNez7FLuDh67l6hlIsWLrmaILwS5azJLYUpobZsxCIc9Ijg5-XShqBEu4lv_rBdVxn4IxQThREdxUZaP0DYOBrqKiv8n7P1FO1W5E7pca-eL-ZHo6btOg?width=1793&height=854&cropmode=none">
</p>
         
