### Here we analyze how the gaps affect the detection of relaxed selection of CYP8B1 by `RELAX` in terms of :
- number of gaps
- length of gaps 
- number of species


### Steps

1. Select the species 

- Afrotheria can be used - since they show strong relaxed selection on CYP8B1 by [Shinde et al., 2019](https://link.springer.com/article/10.1007/s00239-019-09903-6?fbclid=IwAR2UL_uHcWkEfqa1GyJsq95N_t_Lcaq7TOn1UpFVNj-2ikDJnUEbHi0ZBCQ#Sec2). 
    - There were only 7 Afrothrtia species used in Shinde's study.
    - Gaps 'replace' (not 'insert' - which will change reading frame) random nucleotides with N's.
    - Even before introducing gaps - there were total 16 N's already present in the sequence of 2 species.
      - 13 N's in Loxodonta africana (3 triplets + 1 nt)
      - 3 N's in Tricheus manatus latirostris (1 triplet)
    - These N's are also gaps - but are introduced before ORF validation to facilitate alignment.
    - Hence these N's are considered normal w.r.t gaps N's that we introduce in this step.

2. Introduce the gaps

- To introduce gaps we used the `bedtools`.

  - Calculate gene length of all species 
`bioawk -cfastx '{print $name, length($seq)}' afrotheria/orf.fa > afrotheria/gene_length.txt`

  - Make the intervals - of where to introduce the gaps
`bedtools random -n 3 -l 5 -g afrotheria/gene_length.txt > afrotheria/l5_n3_gaps.bed`

  - Introduce the gaps using the intervals made
`bedtools maskfasta -fi afrotheria/orf.fa -bed afrotheria/l5_n3_gaps.bed -fo afrotheria/masked.fa`

  - Run MSA using Guidance 

```perl $g --program GUIDANCE --seqFile masked.fa --msaProgram MAFFT --seqType codon --outDir mafft/ --genCode 1 --bootstraps 100 --proc_num 8
perl $g --program GUIDANCE --seqFile masked.fa --msaProgram PRANK --seqType codon --outDir prank/ --genCode 1 --bootstraps 100 --proc_num 8
perl $g --program GUIDANCE --seqFile masked.fa --msaProgram CLUSTALW --seqType codon --outDir clustalw2 --genCode 1 --bootstraps 100 --proc_num 8
```
