### Objective - To analyze how the detection of relaxed selection of CYP8B1 gene by `RELAX` is affected by the following :
- number of gaps
- length of gaps 
- number of species


### Steps - MSA's and Tree Building

- **1. Select the order and number of species to be used.**  
  
   <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> The order Afrotheria can be used - because they show strong relaxed selection on CYP8B1 gene in Shinde 2019.    :point_right: Points to note : </summary>

  ---

    - There were only 7 Afrothrtia species used in Shinde et al.,'s study.
    - Gaps 'replace' random nucleotides with N's (not 'insert' - which will change reading frame).
    - Even before introducing gaps - there were total 16 N's already present in the sequence of 2 species.
      - 13 N's in Loxodonta africana (4 triplets + 1 nt)
      - 3 N's in Tricheus manatus latirostris (1 triplet)
    - These N's are also gaps - but are introduced before ORF validation to facilitate alignment.
    - Hence these N's are considered normal w.r.t **gap N's** that we introduce in this step.
     
    ---  
  
     </details>
     <!-- AUTO-GENERATED-CONTENT:END -->

---

- **2. Introduce the gaps using bedtools**

  - Calculate gene length of all species : output of this step is the input for the next step - `bedtools random`.
  
        cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
        bioawk -cfastx '{print $name, length($seq)}' afrotheria.fa > gene_length.txt

  - Make the intervals of where to introduce the gaps :
    - n = number of gaps.
    - l = length of gaps.
  
          bedtools random -n 3 -l 5 -g gene_length.txt -seed 666 > l5_n3_gaps.bed
 
  - Introduce the gaps using the intervals made in previous step :

        bedtools maskfasta -fi afrotheria.fa -bed l5_n3_gaps.bed -fo masked.fa

---

- **3. Run MSA using Guidance** 

  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> :point_right: What is Guidance & Why to use it ?</summary>
         
   ---      
         
   - GUIDANCE is  used for weighting, filtering or masking unreliably aligned positions in sequence alignments before subsequent analysis.
     
     <div style="float: right; margin-left: 30px;"><img title="How Guidance works" style="float: right;margin-left: 30px;" src="http://guidance.tau.ac.il/ver2/fig_method.server-1.png" align=right height=320/></div>
     
   - It points to sequences that are unreliably aligned and enables their automatic removal from the MSA, in preparation for downstream analyses.
   - GUIDANCE cannot tell you which alignment is better.
   - GUIDANCE measures the robustness of the alignment and not accuracy.
   
      - So a perfect score means - the aligner will always consistently aligns the sequences in the same way, regardless of perturbations in GUIDANCE.
      - Also a higher score doesn't mean higher accurcy.
  
  ---
  
  </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
  
      cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
      i=/home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf/afrotheria.fa
      guidance --program GUIDANCE --seqFile "$i" --msaProgram PRANK --seqType codon --outDir "$i".100_PRANK --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram MUSCLE --seqType codon --outDir "$i".100_MUSCLE --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram CLUSTALW --seqType codon --outDir "$i".100_CLUSTALW --genCode 1 --bootstraps 100 --proc_num 8
      guidance --program GUIDANCE --seqFile "$i" --msaProgram MAFFT --seqType codon --outDir "$i".100_MAFFT --genCode 1 --bootstraps 100 --proc_num 8
 
 
  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary> :point_right: Error log : </summary>

     ---

  - Running `Guidance` with each aligner throwed this error :
     
     <p align="center">
     <img width="760" height="160" src="https://iqhvlw.dm.files.1drv.com/y4m0tUgOy2fNYW7OR7XZ5JxwFxJL3FbtZMvox4VR2UP5deRgZYnLjuWSvF99ETN3vrWTLxziERpOYKHu6rcSeXlYenSrEot-ZisC_UZ37e23gadCzJDgdOcAL4j3oBg_aVzlSoWGGSh-JstZgVR1PKquVZoCyzC4hsHiBVzBWR1FVSRBpa7p3FNvQWfXUngHPkeBI_6dutBJhd6Lqnsuo42WA?width=950&height=221&cropmode=none">
      </p>
  
  - The command worked fine even with this error.
  - This is not really an error that will affect the results. 
  - It might be some bug in the program.
  - However to confirm check one of the output it made - `Seqs.Codes` 
    - `Seq.Codes` is just list of sequences.
    - Make sure no sequences are repeated.      
      
   ---
      
   </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
      
--- 
 
- **4. Sort MSA files before comparing - Using github repository `seqkit` of shensei356.**
   - Sorting is done because - MSA comparison have to run in the same order in which the species are arranged in MSA.
  
         seqkit sort "$i".100_PRANK/MSA.PRANK.aln.With_Names >"$i"_PRANK.aln
         seqkit sort "$i".100_MUSCLE/MSA.MUSCLE.aln.With_Names >"$i"_MUSCLE.aln
         seqkit sort "$i".100_CLUSTALW/MSA.CLUSTALW.aln.With_Names >"$i"_CLUSTALW.aln
         seqkit sort "$i".100_MAFFT/MSA.MAFFT.aln.With_Names >"$i"_MAFFT.aln
        
        
    <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
  <details>
  <summary> :point_right: Seqkit Output : </summary>

    ---

     <div style="float: right; margin-left: 30px;"><img title="Logo by @gregcaporaso." style="float: right;margin-left: 30px;" src="https://hkjwcw.dm.files.1drv.com/y4ml8T7vybLkXBQovFtyNmoXdoOA-0QH84O_kshlmiMDeMX_ejw3UdPx3R7P5NrqkFmR1mIhifKkd57PtGAn6O_AvHVtYotU52jBlp_OhfzMasQZKVZnEUkJ5_DOoZMEY8uIAxROsXiMds_lqG76CO0-OqPG30hoEImKetfTmWi7XP1fh1YvAzsTQV7auPbodIA4XYuh5gYfpY-aBTKNtzzcQ?width=928&height=327&cropmode=none" align=left height=220/></div>     

 
    &nbsp;
  
    - Above : MSA before sorting.
  
  &nbsp;
  
  &nbsp;
  
    - Below : MSA after sorting.
      
    &nbsp;
    
    &nbsp;

    
  ---
  
   </details>
  <!-- AUTO-GENERATED-CONTENT:END -->
  
---  
  
