### Here we analyze how the detection of relaxed selection of CYP8B1 by `RELAX` is affected by the following :
- number of gaps
- length of gaps 
- number of species


### Steps

- Select the order and number of species to be used.  
  
   <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> The order Afrotheria can be used - because they show strong relaxed selection on CYP8B1 gene in Shinde 2019.    :point_right: Points to note </summary>
  
    - There were only 7 Afrothrtia species used in Shinde's study.
    - Gaps 'replace' (not 'insert' - which will change reading frame) random nucleotides with N's.
    - Even before introducing gaps - there were total 16 N's already present in the sequence of 2 species.
      - 13 N's in Loxodonta africana (3 triplets + 1 nt)
      - 3 N's in Tricheus manatus latirostris (1 triplet)
    - These N's are also gaps - but are introduced before ORF validation to facilitate alignment.
    - Hence these N's are considered normal w.r.t gaps N's that we introduce in this step.
     
  
     </details>
     <!-- AUTO-GENERATED-CONTENT:END -->


- Introduce the gaps using bedtools

  - Calculate gene length of all species :
  
        cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
        bioawk -cfastx '{print $name, length($seq)}' afrotheria.fa > gene_length.txt

  - Make the intervals of where to introduce the gaps :
  
        bedtools random -n 3 -l 5 -g gene_length.txt -seed 666 > l5_n3_gaps.bed
 
  - Introduce the gaps using the intervals made in previous step :

        bedtools maskfasta -fi afrotheria.fa -bed l5_n3_gaps.bed -fo masked.fa


- Run MSA using Guidance 

  <!-- AUTO-GENERATED-CONTENT:START (TOC:collapse=true&collapseText="Click to expand") -->
   <details>
   <summary> :point_right: What is Guidance & Why to use it ?</summary>
       
   - GUIDANCE is  used for weighting, filtering or masking unreliably aligned positions in sequence alignments before subsequent analysis.
   - It points to columns and sequences that are unreliably aligned and enables their automatic removal from the MSA, in preparation for downstream analyses.
   - GUIDANCE cannot tell you which alignment is better.
   - GUIDANCE measures the robustness of the alignment and not accuracy.
   
      - So a perfect score means - the aligner will always consistently aligns the sequences in the same way, regardless of perturbations in GUIDANCE.
      - Also a higher score doesn't mean higher accurcy.
   
   &nbsp;
   
   <p align="center">
     <img width="460" height="300" src="http://guidance.tau.ac.il/ver2/fig_method.server-1.png">
   </p>
   
   
  
  </details>
  <!-- AUTO-GENERATED-CONTENT:END -->

  
  
          cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
          i=/home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf/afrotheria.fa
          guidance --program GUIDANCE --seqFile "$i" --msaProgram PRANK --seqType codon --outDir "$i".100_PRANK --genCode 1 --bootstraps 100 --proc_num 8
          guidance --program GUIDANCE --seqFile "$i" --msaProgram MUSCLE --seqType codon --outDir "$i".100_MUSCLE --genCode 1 --bootstraps 100 --proc_num 8
          guidance --program GUIDANCE --seqFile "$i" --msaProgram CLUSTALW --seqType codon --outDir "$i".100_CLUSTALW --genCode 1 --bootstraps 100 --proc_num 8
          guidance --program GUIDANCE --seqFile "$i" --msaProgram MAFFT --seqType codon --outDir "$i".100_MAFFT --genCode 1 --bootstraps 100 --proc_num 8
 
- Sort MSA files before comparing - Using github repository `seqkit` of shensei356.
  
      seqkit sort "$i".100_PRANK/MSA.PRANK.aln.With_Names >"$i"_PRANK.aln
      seqkit sort "$i".100_MUSCLE/MSA.MUSCLE.aln.With_Names >"$i"_MUSCLE.aln
      seqkit sort "$i".100_CLUSTALW/MSA.CLUSTALW.aln.With_Names >"$i"_CLUSTALW.aln
      seqkit sort "$i".100_MAFFT/MSA.MAFFT.aln.With_Names >"$i"_MAFFT.aln
        
- Compare alignments to get AOS and MOS - Using `mumsa` from [msa.sbc.su.se](http://msa.sbc.su.se/)

      echo "MUMSA comparison results of mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln" > "$i".log
      mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln >> "$i".log
      alncomp=`mumsa -g -s "$i"_PRANK.aln "$i"_MUSCLE.aln "$i"_CLUSTALW.aln "$i"_MAFFT.aln`
      
- Perform model testing on each alignment - Using github repository `modeltest` of ddarriba.

      modeltest-ng -i "$i"_MAFFT.aln -t ml -T raxml
      modeltest-ng -i "$i"_PRANK.aln -t ml -T raxml
      modeltest-ng -i "$i"_MUSCLE.aln -t ml -T raxml
      modeltest-ng -i "$i"_CLUSTALW.aln -t ml -T raxml
      
      for j in "$i"_PRANK "$i"_MUSCLE "$i"_CLUSTALW "$i"_MAFFT
      do
      BICmodel=`grep -A 2 "Best model according to BIC" "$j".aln.out|tail -1|awk '{print $2}'`
      AICmodel=`grep -A 2 "Best model according to AICc" "$j".aln.out|tail -1|awk '{print $2}'`
      echo $j $BICmodel $AICmodel
      done

      j="$i"_PRANK
      model=`grep -A 2 "Best model according to BIC" "$j".aln.out|tail -1|awk '{print $2}'`
 
- Using the PRANK alignment best model according to BIC
       
      echo "Model $model has been selected." >> "$i".log

- Build phylogenies with bootstrap - Using github repository `raxmlng` of amkozlov

      raxml-ng --all -msa "$i"_PRANK.aln --model $model --bs-trees 1000 --threads 2
      roottax=`grep ">" "$i"_PRANK.aln|head -1|sed 's/>//g'`
