### RELAX ON CYP8B1 gene in 7 Afrotheria species.

Here is the script of running RELAX on CYP8B1 gene with 7 Afrotheria species. Each species is taken as a forground (fg) species with rest of the 6 species as background (bg).
Here is the wrkflow :

Download ORFs -> Introduce Gaps -> Run MSA -> Sort MSA -> Compare MSA -> Find best model -> Infer best tree -> Label fg and bg -> Make config file for HYPHY -> Run HYPHY -> Results

#### 1. Download ORFs

    cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
    wget https://raw.githubusercontent.com/ceglab/CYP8B1/master/ORFs/2_Afrotheria.fa
    mv 2_Afrotheria.fa afrotheria.fa

#### 2. Introduce Gaps
    cd /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/gaps/afrotheria/orf
    bioawk -cfastx '{print $name, length($seq)}' afrotheria.fa > gene_length.txt
    bedtools random -n 3 -l 5 -g gene_length.txt -seed 666 > l5_n3_gaps.bed
    bedtools maskfasta -fi afrotheria.fa -bed l5_n3_gaps.bed -fo masked.fa

#### 3. Run MSA with guidance
    perl /home/ceglab8/workspace/phd/research/efficiency_of_RELAX/CYP8B1/tools/guidance.v2.02/www/Guidance/guidance.pl --program
    GUIDANCE --seqFile masked.fa --msaProgram PRANK --seqType codon --outDir prank_100 --genCode 1 --bootstraps 100 --proc_num 8
    seqkit sort prank_100/MSA.PRANK.aln.With_Names > CDM_PRANK.aln

#### 4. Find best model.
    modeltest-ng -i CDM_PRANK.aln -t ml -T raxml

#### 5. Infer best tree
    raxml-ng --all --msa CDM_PRANK.aln --model HKY+G4 --bs-trees 1000 --threads 2

    

#### 6. Label tree

    t=`grep ">" CDM_PRANK.aln|wc -l`
    grep ">" CDM_PRANK.aln|sed 's/>//g' > taxlist.txt
    
    for i in `cat taxlist.txt`
    do sed "s/$i/$i{fg}/g" CDM_PRANK.aln.raxml.bestTree > "$i"_treeLabled.txt
    done

#### 7. make config file for HYPHY
    for i in `cat taxlist.txt`
    do echo -ne  "14\n9\n1\n"$PWD"/CDM_PRANK.aln\n"$PWD"/"$i"_treeLabled.txt\n2\n2" > "$i"_tree.config
    done

#### 8. Running Hyphy minimal
    for i in `cat taxlist.txt`
    do HYPHYMP<"$i"_tree.config > "$i"_treeoutput_HYPHY
    done

#### 9. Check the results

    for i in $(ls *_HYPHY);do echo $i; grep "Relaxation/intensification parameter (K)" $i; tail -4 $i ;done

    for i in $(ls *_HYPHY);do grep "Relaxation/intensification parameter (K)" $i ;done
    for i in $(ls *_HYPHY);do grep "Likelihood ratio test" $i ;done
